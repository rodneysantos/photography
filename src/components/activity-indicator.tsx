import clonedeep from "lodash.clonedeep";
import lottie, { AnimationItem } from "lottie-web";
import { createSignal, onCleanup, onMount } from "solid-js";
import animationData from "../assets/camera-aperture.json";

export const ActivityIndicator = () => {
	const [animationRef, setAnimationRef] = createSignal<AnimationItem | null>(
		null,
	);

	onCleanup(() => {
		// Cleanup code, stop the animation if necessary
		animationRef()?.destroy();
	});

	onMount(() => {
		const container = document.getElementById("container") as HTMLElement;
		const animation = lottie.loadAnimation({
			container,
			renderer: "svg",
			loop: true,
			autoplay: true,
			animationData: clonedeep(animationData),
		});

		animation.setSpeed(2.5);
		setAnimationRef(animation);
	});

	return <div id="container" class="w-56"></div>;
};
