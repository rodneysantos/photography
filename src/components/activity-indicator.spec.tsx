import { render } from "@solidjs/testing-library";
import lottie, { AnimationItem } from "lottie-web";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { ActivityIndicator } from "./activity-indicator";

vi.mock("lottie-web");

describe("activity-indicator", () => {
	const destroyMock = vi.fn();
	const setSpeedMock = vi.fn();

	beforeEach(() => {
		vi.spyOn(lottie, "loadAnimation").mockReturnValue({
			setSpeed: setSpeedMock,
			destroy: destroyMock,
		} as unknown as AnimationItem);
	});

	it("renders correctly", () => {
		// assemble
		const { asFragment } = render(() => <ActivityIndicator />);

		// assert
		expect(asFragment()).toMatchSnapshot();
	});

	it("sets the speed on mount", () => {
		// assemble
		render(() => <ActivityIndicator />);

		// assert
		expect(setSpeedMock).toHaveBeenCalledWith(2.5);
	});

	it("destroys the animation on cleanup", () => {
		// assemble
		const { unmount } = render(() => <ActivityIndicator />);

		// act
		unmount();

		// assert
		expect(destroyMock).toHaveBeenCalled();
	});
});
