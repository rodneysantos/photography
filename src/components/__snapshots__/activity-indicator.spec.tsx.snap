// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`activity-indicator > renders correctly 1`] = `"<div id=\\"container\\" class=\\"w-56\\"></div>"`;
